File: sdks/go/devenv_tunnel.go
// Package devenvtunnel provides a thin helper for devenv-tunnel integration.
//
// The universal mechanism (no real SDK needed):
//  1. Set DEVENV_TUNNEL to a full domain name BEFORE starting your process.
//     The suffix decides the route:
//       - myapp-{branch}.devenv.local        → local virtual overlay
//       - myapp-{branch}.tunnel.devenv.tools → cloud tunnel
//     See sdks/direnv/README.md for the recommended direnv setup.
//  2. Bind your server to port 0 — the OS assigns an ephemeral port.
//  3. The devenv-tunnel daemon discovers the process, reads DEVENV_TUNNEL,
//     finds the real port, and routes traffic to it.
//
// IMPORTANT — environment visibility:
// The daemon reads each process's environment from OUTSIDE the process:
//   - Linux: /proc/<pid>/environ
//   - macOS: sysctl KERN_PROCARGS2
//
// Both sources are frozen snapshots from execve() time. Calling os.Setenv
// at runtime updates only the in-process copy and is NEVER visible to the
// daemon. This package does NOT set DEVENV_TUNNEL — set it before launch.
//
// This package has no external dependencies.
package devenvtunnel

import (
	"fmt"
	"log"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// Options controls how FindFreeListener and FindFreePort behave.
type Options struct {
	// Host is the interface to bind on. Defaults to "0.0.0.0".
	Host string

	// ServiceName is used in log output. Defaults to "app".
	ServiceName string

	// Logger is used for output. Defaults to log.Printf when nil.
	Logger func(format string, args ...any)
}

func (o *Options) host() string {
	if o.Host == "" {
		return "0.0.0.0"
	}
	return o.Host
}

func (o *Options) serviceName() string {
	if o.ServiceName == "" {
		return "app"
	}
	return o.ServiceName
}

func (o *Options) logf(format string, args ...any) {
	if o.Logger != nil {
		o.Logger(format, args...)
	} else {
		log.Printf(format, args...)
	}
}

// resolveTemplateForDisplay attempts to substitute {branch} and {worktree}
// placeholders using local git context. This is for display/logging purposes
// only — the daemon resolves templates independently from cwd/git context.
// Failure is non-fatal; unresolved placeholders remain as-is.
func resolveTemplateForDisplay(template string) (resolved string, changed bool) {
	if !strings.Contains(template, "{branch}") && !strings.Contains(template, "{worktree}") {
		return template, false
	}

	resolved = template

	if branch, err := gitOutput("rev-parse", "--abbrev-ref", "HEAD"); err == nil && branch != "" {
		resolved = strings.ReplaceAll(resolved, "{branch}", branch)
	}

	if root, err := gitOutput("rev-parse", "--show-toplevel"); err == nil && root != "" {
		resolved = strings.ReplaceAll(resolved, "{worktree}", filepath.Base(root))
	}

	return resolved, resolved != template
}

func gitOutput(args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Stderr = nil
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

// ReadTunnel returns the value of DEVENV_TUNNEL from the environment, or an
// empty string if unset. This function is read-only — it does not set the
// variable. The daemon reads /proc/<pid>/environ at launch time and cannot
// see runtime os.Setenv changes.
func ReadTunnel() string {
	return os.Getenv("DEVENV_TUNNEL")
}

// FindFreeListener binds a TCP listener to port 0 and returns it along with
// the assigned port number. The listener stays open — pass it to your HTTP
// server (e.g. http.Serve(ln, handler)) or close it yourself.
//
// DEVENV_TUNNEL must already be set before the process was started.
// If unset, a WARNING is logged with setup instructions.
func FindFreeListener(opts Options) (net.Listener, int, error) {
	tunnel := ReadTunnel()

	addr := fmt.Sprintf("%s:0", opts.host())
	ln, err := net.Listen("tcp", addr)
	if err != nil {
		return nil, 0, fmt.Errorf("devenv-tunnel: failed to listen on %s: %w", addr, err)
	}

	tcpAddr, ok := ln.Addr().(*net.TCPAddr)
	if !ok {
		ln.Close() //nolint:errcheck
		return nil, 0, fmt.Errorf("devenv-tunnel: unexpected address type %T", ln.Addr())
	}
	port := tcpAddr.Port

	if tunnel != "" {
		display, changed := resolveTemplateForDisplay(tunnel)
		note := ""
		if changed {
			note = " (informational local resolution — daemon resolves independently)"
		}
		opts.logf("[devenv-tunnel] %s bound to %s:%d — tunnel domain: %s%s",
			opts.serviceName(), opts.host(), port, display, note)
	} else {
		opts.logf("[devenv-tunnel] WARNING: %s bound to %s:%d — DEVENV_TUNNEL is not set.\n"+
			"  The daemon reads the process environment at launch time (execve snapshot) and\n"+
			"  cannot see runtime os.Setenv changes.\n"+
			"  Set DEVENV_TUNNEL before starting your process:\n"+
			"    direnv:  export DEVENV_TUNNEL=myapp-$(git rev-parse --abbrev-ref HEAD).devenv.local\n"+
			"    shell:   export DEVENV_TUNNEL=myapp-{branch}.devenv.local\n"+
			"    docker:  docker run -e DEVENV_TUNNEL=myapp-{branch}.devenv.local ...\n"+
			"  See sdks/direnv/README.md for the recommended setup.",
			opts.serviceName(), opts.host(), port)
	}

	return ln, port, nil
}

// FindFreePort binds to port 0, records the port number, closes the listener,
// and returns the port. Use only when your framework requires a port integer
// rather than a net.Listener. There is a brief TOCTOU window — prefer
// FindFreeListener when possible.
func FindFreePort(opts Options) (int, error) {
	ln, port, err := FindFreeListener(opts)
	if err != nil {
		return 0, err
	}
	ln.Close() //nolint:errcheck
	return port, nil
}

File: sdks/python/devenv_tunnel.py
"""
devenv_tunnel — thin Python helper for devenv-tunnel.

The universal mechanism (no real SDK needed):
  1. Set DEVENV_TUNNEL to a full domain name BEFORE starting your process.
     The suffix decides the route:
       - myapp-{branch}.devenv.local       → local virtual overlay
       - myapp-{branch}.tunnel.devenv.tools → cloud tunnel
     See sdks/direnv/README.md for the recommended direnv setup.
  2. Bind your server to port 0 — the OS assigns an ephemeral port.
  3. The devenv-tunnel daemon discovers the process, reads DEVENV_TUNNEL,
     finds the real port, and routes traffic to it.

IMPORTANT — environment visibility:
  The daemon reads each process's environment from OUTSIDE the process
  (/proc/<pid>/environ on Linux, sysctl KERN_PROCARGS2 on macOS). Both
  sources are frozen snapshots from execve() time. Setting
  os.environ["DEVENV_TUNNEL"] at runtime updates only the in-process
  libc copy and is NEVER visible to the daemon.

  This module does NOT and CANNOT set DEVENV_TUNNEL for daemon discovery.
  Set it before starting your process (direnv / shell export / docker -e).

This module is stdlib-only (no pip install needed).
"""

from __future__ import annotations

import logging
import os
import socket
import subprocess
from typing import Optional

logger = logging.getLogger(__name__)


def _resolve_template_for_display(template: str) -> str:
    """Resolve {branch}/{worktree} placeholders locally for logging only.

    The daemon resolves these independently from cwd/git context — this
    local resolution is purely informational. Failure is non-fatal.
    """
    if "{branch}" not in template and "{worktree}" not in template:
        return template

    branch: Optional[str] = None
    worktree: Optional[str] = None

    try:
        branch = subprocess.check_output(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            stderr=subprocess.DEVNULL,
            timeout=2,
        ).decode().strip()
    except Exception:
        pass  # Non-fatal

    try:
        worktree_root = subprocess.check_output(
            ["git", "rev-parse", "--show-toplevel"],
            stderr=subprocess.DEVNULL,
            timeout=2,
        ).decode().strip()
        worktree = os.path.basename(worktree_root)
    except Exception:
        pass  # Non-fatal

    resolved = template
    if branch:
        resolved = resolved.replace("{branch}", branch)
    if worktree:
        resolved = resolved.replace("{worktree}", worktree)
    return resolved


def read_tunnel() -> Optional[str]:
    """Read DEVENV_TUNNEL from the environment.

    Returns the value, or None if unset.  Does NOT set the variable —
    the daemon reads /proc/<pid>/environ at launch time and cannot see
    runtime os.environ changes.
    """
    return os.environ.get("DEVENV_TUNNEL")


def find_free_port(
    host: str = "0.0.0.0",
    *,
    service_name: str = "app",
    log: bool = True,
) -> tuple[socket.socket, int]:
    """Bind a TCP socket to port 0 and return ``(sock, port)``.

    The caller is responsible for closing the socket (or passing it to a
    framework that takes ownership of it).

    DEVENV_TUNNEL must already be set in the environment before this
    process was started.  If it is unset, a WARNING is logged with setup
    instructions.

    Parameters
    ----------
    host:
        Interface to bind on. Defaults to all interfaces.
    service_name:
        Label used in log output.
    log:
        Emit an INFO log line (or WARNING when DEVENV_TUNNEL is unset).
        Set to False to suppress output.

    Returns
    -------
    (sock, port)
        ``sock`` is already bound; pass it to your server framework or close
        it after recording the port.
    """
    tunnel = read_tunnel()

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind((host, 0))
    _, port = sock.getsockname()

    if log:
        if tunnel:
            display = _resolve_template_for_display(tunnel)
            note = (
                " (informational local resolution — daemon resolves independently)"
                if display != tunnel
                else ""
            )
            logger.info(
                "[devenv-tunnel] %s bound to %s:%d — tunnel domain: %s%s",
                service_name,
                host,
                port,
                display,
                note,
            )
        else:
            logger.warning(
                "[devenv-tunnel] WARNING: %s bound to %s:%d — DEVENV_TUNNEL is not set.\n"
                "  The daemon reads the process environment at launch time and cannot see "
                "runtime os.environ changes.\n"
                "  Set DEVENV_TUNNEL before starting your process:\n"
                "    direnv:  add export DEVENV_TUNNEL=myapp-$(git rev-parse --abbrev-ref HEAD)"
                ".devenv.local  to .envrc\n"
                "    shell:   export DEVENV_TUNNEL=myapp-{branch}.devenv.local\n"
                "    docker:  docker run -e DEVENV_TUNNEL=myapp-{branch}.devenv.local ...\n"
                "  See sdks/direnv/README.md for the recommended setup.",
                service_name,
                host,
                port,
            )

    return sock, port


def get_free_port(
    host: str = "0.0.0.0",
    *,
    service_name: str = "app",
    log: bool = True,
) -> int:
    """Return an ephemeral port number (the bound socket is closed immediately).

    Use this when you need the port *number* before constructing your server
    and the framework accepts a port integer rather than a pre-bound socket.

    Note: there is a brief TOCTOU window between closing the socket and the
    server rebinding; ``find_free_port`` (which keeps the socket open) is
    safer when the framework supports it.
    """
    sock, port = find_free_port(host, service_name=service_name, log=log)
    sock.close()
    return port
